src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs

```diff
@@ -515,14 +532,30 @@ public Task RelayTransaction(uint256 hash)
             // find all behaviours then start an exclusive task
             // to add the hash to each local collection
             IEnumerable<MempoolBehavior> behaviours = peers.Select(s => s.Behavior<MempoolBehavior>());
-            this.logger.LogTrace("(-)");
-            return this.manager.MempoolLock.WriteAsync(() =>
+            return this.manager.MempoolLock.WriteAsync(() => 
             {
                 foreach (MempoolBehavior mempoolBehavior in behaviours)
                 {
+                    this.logger.LogTrace("Attempting to relaying transaction ID '{0}' to peer '{1}'.", hash, mempoolBehavior?.AttachedPeer.RemoteSocketEndpoint);
                     if (mempoolBehavior?.AttachedPeer.PeerVersion.Relay ?? false)
-                        if (!mempoolBehavior.filterInventoryKnown.ContainsKey(hash))
-                            mempoolBehavior.inventoryTxToSend.TryAdd(hash, hash);
+                    {
```

Comment:
All the code form this ` if (mempoolBehavior?.AttachedPeer.PeerVersion.Relay ?? false)` should go in its own private method.
`mempoolBehavior.TryRelay()` then you can use the lock locally.

If this current behaviours does not drop nodes then all is ok.